File: PlantsRPetsProjeto.Server/Services/WeatherService.cs
using Newtonsoft.Json;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

namespace PlantsRPetsProjeto.Server.Services
{
    /// <summary>
    /// Serviço responsável por obter dados meteorológicos de uma localização específica
    /// através da API externa WeatherAPI.
    /// </summary>
    public class WeatherService
    {
        private readonly HttpClient _httpClient;
        private readonly string _apiKey = "1e1bfcd7c4e5498791d120909252602";

        /// <summary>
        /// Inicializa uma nova instância do <see cref="WeatherService"/> com um cliente HTTP configurado.
        /// </summary>
        /// <param name="httpClient">Instância de <see cref="HttpClient"/> utilizada para fazer os pedidos à API externa.</param>
        public WeatherService(HttpClient httpClient)
        {
            _httpClient = httpClient;
        }

        /// <summary>
        /// Obtém os dados meteorológicos para uma dada localização (nome de cidade ou coordenadas).
        /// A previsão devolvida inclui três dias, sem índice de qualidade do ar nem alertas.
        /// </summary>
        /// <param name="location">Localização a consultar (ex: nome da cidade ou coordenadas no formato "lat,lon").</param>
        /// <returns>Objeto com os dados meteorológicos ou null caso a resposta não seja bem-sucedida.</returns>
        public async Task<object> GetWeatherAsync(string location)
        {
            var response = await _httpClient.GetAsync($"http://api.weatherapi.com/v1/forecast.json?key={_apiKey}&q={location}&days=3&aqi=no&alerts=no");

            if (!response.IsSuccessStatusCode)
            {
                return null;
            }

            var weatherData = await response.Content.ReadFromJsonAsync<object>();
            return weatherData;
        }
    }
}

File: plantsrpetsprojeto.client/src/app/weather.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})

/**
* Serviço responsável por obter informações meteorológicas da API.
* Permite obter previsões por nome de cidade ou por coordenadas geográficas.
*/
export class WeatherService {
  constructor(private http: HttpClient) { }

  /**
   * Obtém a previsão do tempo com base no nome da cidade.
   * @param city Nome da cidade (ex: 'Lisboa')
   * @returns Observable com os dados meteorológicos
   */
  getWeatherByCity(city: string): Observable<any> {
    return this.http.get(`/api/weather/city/${city}`);
  }

  /**
   * Obtém a previsão do tempo com base nas coordenadas geográficas.
   * @param lat Latitude
   * @param lon Longitude
   * @returns Observable com os dados meteorológicos
   */
  getWeatherByCoords(lat: number, lon: number): Observable<any> {
    return this.http.get(`/api/weather/coords/${lat}/${lon}`);
  }
}
